### DISCUSSION TITLE
Scaffold Tool Feedback

### DISCUSSION THREAD

**Discussion by MaceWindu** [linq2db/linq2db#3531] [General]
This discussion created to gather feedback for new database scaffolding tool ([`linq2db.cli`](https://www.nuget.org/packages/linq2db.cli)) which replaces T4 templates in linq2db 4.0.

T4 templates still shipped, but we encourage to migrate from them to new tool. We don't plan any changes to templates anymore and will discontinue them in future.

Initial release: 4.0.0-rc.1 (https://github.com/linq2db/linq2db/pull/3098)

RC2 added:
- remaining extension points to customize code generation for entities, functions, procedures and associations
- option to generate `IEquatable<T>` on entity classes (`Equatable.ttinclude` template migration)

**Comment by Powerz** [comment]
As a huge fan of Linq2Db who's been using it since good old Bltoolkit days I want to say thank you for making this world a better place by working on the best .net ORM ever ❤️

This CLI tool is absolutely fantastic, and one of the reasons is that it finally works on macOS unlike the old T4 templates.

Would like to see an example on how to debug it when run with customization (.net class library + ScaffoldInterceptors). To be more precise, how to debug the custom interceptor.

Small bits that require attention:

https://github.com/linq2db/linq2db/issues/3564

```
   --single-file                             : generates all code in single file
         type: bool (optional)
         json: code.single-file
         default: false
```

The default value is `true`, at least I had to explicitly change it to `false` in order to get separate files.

Once again, thank you! @igor-tkachev is a legend, all contributors are rockstars!

**Reply by MaceWindu** [reply]
Thanks! That's really looks like a help bug, it should be `true`.

**Reply by MaceWindu** [reply]
Defaults are correct, but in your case you probably used compatibility mode with old t4 templates (`--template t4`) where it is `true` by default. Next release (4.0.1) will add defaults for this mode too to help: `default (t4 mode): true`

**Reply by AndreasHaackEpi** [reply]
The CLI tool looks very promising and knowing your work, I'm confident that it will be as good as it looks.
I'm particularly interested in the extensibility with an assembly. However, I too have no clue how to debug my assembly (besides perhaps launching the LinqToDB.CLI project itself in Visual Studio).
Is there any other possibility I'm not seeing? And if so, what is it?

Thanks again for your amazing work. From time to time I'm forced to use EF at my job and its limitations are driving me nuts. I'm always happy when I can go back to the projects which are under my aegis and which **Don't' EF**.

**Reply by AndreasHaackEpi** [reply]
Forget it. I saw the answer you gave elsewhere.
`System.Diagnostics.Debugger.Launch()`

**Reply by richinoz** [reply]
Hi AndreasHaackEpi,
Can you explain how you debug your assembly using the tool?

**Reply by AndreasHaackEpi** [reply]
Hi, @richinoz , sorry I saw your messages only now. 
Add System.Diagnostics.Debugger.Launch() to your code and the Debugger is launched

**Comment by Powerz** [comment]
Proposal: add the `generated` suffix to the generated model files e.g. `User.generated.cs` instead of `User.cs`. In this case it'd be easier to make the generated files co-exist with their custom partial brothers.

**Reply by MaceWindu** [reply]
Will be fixed in 4.0.1

**Reply by ademchenko** [reply]
I would also ask to add the ability to customize that suffix, because, for example, we usually use more short version "auto", i.e. User.auto.cs.

**Comment by panayis** [comment]
Can I use the new CLI tool with a full dot net framework 4.8 winforms application? ..or is it only for dot net core?

**Reply by MaceWindu** [reply]
Sure. Tool needs .net core 3.1+ to run, but generated code could be used with any supported by linq2db framework

**Comment by Powerz** [comment]
I've been trying to add a foreign key (which doesn't exist in the database) using a T4 scaffold interceptor, but fail miserably. Need your help :)

1. In the `GetForeignKeys` method I add the FK.

Example: 
1 Order can have 1 or more Items.
Item.OrderId >> Order.Id

```csharp
public class SuperInterceptor : LinqToDB.Scaffold.ScaffoldInterceptors
{
    public override IEnumerable<ForeignKey> GetForeignKeys(IEnumerable<ForeignKey> keys)
    {
        var source = new SqlObjectName("Item", Schema: "dbo");
        var target = new SqlObjectName("Order", Schema: "dbo");
        var relation = new[] { new ForeignKeyColumnMapping("OrderId", "Id") };
        var fk = new ForeignKey("FK_Item_Order", source, target, relation);

        return keys.Union(new[] { fk });
    }
}
```

2. Then I expected `PreprocessAssociation` to get called for this FK where I would be able to specify some details, but it didn't happen, nor the reference was generated.

Could you please advise on how to add an FK using the interceptor?

**Reply by MaceWindu** [reply]
I suspect you need to specify `Database` for source/target or clear `Database` value in Table/View interceptor

**Reply by MaceWindu** [reply]
Will remove dbname by default in next release https://github.com/linq2db/linq2db/issues/3578

**Reply by Powerz** [reply]
It worked! Much obliged.

**Comment by panayis** [comment]
I have used the CLI Tool in a new winForms project and it works indeed very well. However, I need the CLI Tool to generate partial classes instead. How can I manage that?
Many thanks!

**Reply by MaceWindu** [reply]
you can use `--partial-entities true` option

**Reply by panayis** [reply]
This can be added in Database.json?
{
  "general": {
    "provider": "SqlServer",
    "connection": "Data Source=HP\\SQLEXPRESS;Initial Catalog=mydb;Persist Security Info=True;User ID=sa;Password=1234;TrustServerCertificate=True",
    "output": "DataModel",
    "overwrite": true
  },
  "code": {
    "nrt": false
  }
}

**Reply by MaceWindu** [reply]
```json
{
"general": {
"provider": "SqlServer",
"connection": "Data Source=HP\SQLEXPRESS;Initial Catalog=mydb;Persist Security Info=True;User ID=sa;Password=1234;TrustServerCertificate=True",
"output": "DataModel",
"overwrite": true
},
"code": {
"nrt": false
},
// here
"dataModel": {
    "partial-entities": true
}
}
```

**Reply by MaceWindu** [reply]
all cli options (except couple of cli-specific) available in json config

**Reply by panayis** [reply]
Thank you plenty - It works like a charm

**Comment by zzmyers** [comment]
I have used the new CLI and I am working my way through the options and trying to emulate as much as I can of my old T4 output (just to make things simpler). So far, so good. But one thing I am having trouble with is mapping columns in the table to .NET types like so in T4:

Tables["area"].Columns["area_type"].Type = "AreaType";

AreaType is an Enum which maps to the MySQL table which has an ENUM column (which in MySQL is stored as text).

Is there an option to do such a mapping in the CLI and I have just missed it? Or am I out of luck?

**Reply by MaceWindu** [reply]
https://linq2db.github.io/articles/CLI.html#data-model-interceptors

Check first example in `PreprocessEntity` method if you want to replace type of specific column.

**Reply by zzmyers** [reply]
Thanks. I see that now and that makes sense. I wasn't looking at the Interceptors.

**Comment by angelMachin** [comment]
I have been using this tool with a PostgreSQL db and it works very well.
The only issue I have found is the support of partitioned tables. 

The tool creates an empty class for the master table and a set of classes with all the columns for each partition. 
I have a "Sample"  table partitioned by a date column and it creates something like this:

---------------------------- Sample.cs ----------------------------
-- master table

```c#
[Table("sample")]
public class Sample
{
}
```

---------------------------- Sample201801.cs ----------------------------

-- partition for January 2018 

```c#
[Table("sample_2018_01")]
public class Sample201801
{
	[Column("serial_number", CanBeNull=false, IsPrimaryKey = true, PrimaryKeyOrder = 0)] 
        public string    SerialNumber  { get; set; } = null!; // character varying(50)
	[Column("date_time", IsPrimaryKey = true ,PrimaryKeyOrder = 1)] 
        public DateTime DateTime { get; set; } // timestamp (6) without time zone
	[Column("value")] 
        public float Value { get; set; } // real
}
```

-----------------------------------------------------------------------------------


Normally, partitions are not managed from the ORM, it is a DB implementation detail. 
I think the best approach would be to build a class for the master table with all the columns and, eventually, manage the partitions as subclasses:

[Table("sample")]
public class Sample
{
	[Column("serial_number", CanBeNull=false, IsPrimaryKey = true, PrimaryKeyOrder = 0)] 
        public string    SerialNumber  { get; set; } = null!; // character varying(50)
	[Column("date_time", IsPrimaryKey = true ,PrimaryKeyOrder = 1)] 
        public DateTime DateTime { get; set; } // timestamp (6) without time zone
	[Column("value")] 
        public float Value { get; set; } // real
}

[Table("sample_2018_01")]
public class Sample201801 : Sample
{ }

[Table("sample_2018_02")]
public class Sample201802 : Sample
{ }

**Reply by MaceWindu** [reply]
@angelMachin I've created separate issue #3599 to track it

**Comment by mariorancic01** [comment]
i am trying to do custom scaffolding with this command
dotnet linq2db scaffold -i database.json --customize bin\Debug\net6.0\MyData01.dll
but I am getting this exception:
Could not load file or assembly 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

**Reply by sdanyliv** [reply]
Your customisation project should be targeted to `netstandard2.0`.

**Reply by mariorancic01** [reply]
vow, thx for such a quick response. this worked.

**Reply by MaceWindu** [reply]
netstandard2.1 / netcoreapp3.1 will also work (in other words anything compatible with .net core 3.1, which used to run tool)

**Reply by mariorancic01** [reply]
ok thx, but should it work in .net 6 because it didnt work for me

**Reply by MaceWindu** [reply]
it should work in .net 3.1 or greater including .net 6.0, but custom dll should be built with target framework, supported by .net 3.1 (e.g. netstandard2.0, netstandard2.1 or netcoreapp3.1)

**Comment by Mitchies1970** [comment]
**Stored Procedure in SQL Server**
For some reason using a JSON file as per below, the SPs are not being generated, as in not coming through in the generated C# code. I have a simple test DB to get my JSON file right but as yet have been unable to get the SPs to be generated. 

Below is my SP just for the sake of clarity, very simple one. 
`Create   Procedure [Company].[GetPeopleInBranch]
@BranchId int
AS
Select p.*, b.[Name] as BranchName
from People as p 
inner join Company.BranchPeople as bp 
on p.ID = bp.PersonId
Inner Join Company.Branches as b
on b.BranchId = bp.PersonId
Where b.BranchId = @BranchId`

Any help would be very much appreciated.

**Reply by Mitchies1970** [reply]
Below is my JSON file. 

<img width="754" alt="image" src="https://user-images.githubusercontent.com/13706088/201240577-e011bd25-bb16-471b-b074-4521fcef8cc2.png">

**Reply by MaceWindu** [reply]
You need to specify loaded schema objects (schema.objects option):
```json
{ "schema": { "objects": ["table", "view", "foreign-key", "stored-procedure"] } }
```
by default it loads only tables, views and FKs

**Comment by ademchenko** [comment]
I have a generated column in my postgres database table. For some technical reasons it's always null and I wouldn't want to have it in my generated classes. What is the best way to achieve it?

**Reply by MaceWindu** [reply]
https://linq2db.github.io/articles/CLI.html#schema-load-interceptors

You can modify table columns in `GetTables` interceptor

**Comment by richinoz** [comment]
Thanks Andreas. I got it going in the end. 👍

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by MaceWindu** [comment]
Closing it, please use separate issue/discussions for specific questions
